enhancements/20220822-storage-network/20220822-storage-network.md

```diff
@@ -0,0 +1,134 @@
+
+VM and Storage Network will have different VLAN ID.
+
+#### Story 1
+
+VM and Storage Network will use the same interfaces but different VLAN IDs.
+
+- VM and Storage want to share eth2 and eth3.
+- Create VlanConfig with eth2 and eth3
+- Create Network-Attatchment-Definition for VM
+- Create Network-Attatchment-Definition with Whereabouts config for Storage
+
+#### Story 2
+
+VM and Storage Network will use different interfaces.
+
+- VM use eth2 and eth3
+- Storage uses eth4 and eth5
+- Create a new VlanConfig with eth4 and eth5
+- Create Network-Attatchment-Definition with Whereabouts config for Storage only
+
+### User Experience In Detail
+
+#### Manual
+
+1. Users have to shutdown all VMs and stop all Pods attached with Longhorn volumes manually before setting Longhorn Storage Network. (Including Grafana and Prometheus)
+2. Users need to assign a bridge from VLAN Config and a VLAN ID for the storage network.
+3. Setting Longhorn setting for storage network.
+4. Restart pods and VMs manually.
+
+#### Controller
+
+1. Users have to shotdown all VMs and stop all VMI before setting `storage-network` in Harvester settings
+2. Users need to assign a bridge from VLAN Config and a VLAN ID for storage network (Need to setup IPAM Whereabouts IP range)
+3. Controller will shutdown management pods attached with Longhorn volumes and check all volumes are detached.
+4. Controller will set `storage-network` config to Longhorn settings `storage-network`.
+5. Controller will restart management pods.
+6. Users need to restart VM manually.
+
+### API changes
+
+- New Settings `storage-network`
+
+## Design
+
+### Implementation Overview
+
+- Re-use VLAN Bridge from VLAN enhancement
+- Introduce IPAM [Whereabouts](https://github.com/k8snetworkplumbingwg/whereabouts) to be a global IPAM for Longhorn Pods.
+- Create a Network Attachment Definition with plugin bridge and Whereabouts IPAM
+
+```json
+{
+    "cniVersion":"0.3.1",
+    "name":"v1",
+    "type":"bridge",
+    "bridge":"BRIDGE_NAME",
+    "vlan":VLAN_ID,
+    "ipam":{
+        "type":"whereabouts",
+        "range":"192.168.0.0/24"
+    }
+}
+```
+
+#### Phase 2
+
+Functions in this section will not be implemented in v1.1.0
+
+- add Webhook to check VM status while changing `storage-network` settings
+- add Webhook to verify NAD status while changing `storage-network` settings
+
+### Test plan
+
+1. Shutdown all VMs
+2. install whereabouts charts https://github.com/harvester/harvester/pull/2716
+3. setup VLAN with bridge name and IPAM whereabouts config
+```json
+{
+    "cniVersion":"0.3.1",
+    "name":"v1",
+    "type":"bridge",
+    "bridge":"BRIDGE_NAME",
+    "vlan":VLAN_ID,
+    "ipam":{
+        "type":"whereabouts",
+        "range":"192.168.0.0/24"
+    }
+}
+```
+
+4. add Settings named "storage-network" with NAD name in Harvester settings
+5. wait and check Longhorn settings "storage-network" is synced.
+6. check Longhorn `instance-manager-e-XXXX` has an interface named "lhnet1"
```

Comment:
Say the `Controller` is the recommended solution to set `stoarge network`, then `Test plan` needs to be aligned with `Controller`.
When will the `Controller` shutdown internal PODs, when will start again. By which flag/state, the user can start VM again.

Reply:
fixed